### PR TITLE
Accept a URL parameter

### PR DESCRIPTION
**Pull request by iamandrewluca**
This will allow to check any url for GitHub URLs

e.g

```js
isRepoDiscussionList(new URL(document.referref))
isRepoDiscussionList(window.location)
```

Blocks https://github.com/sindresorhus/refined-github/pull/2974

### CODE REVIEW

**Comment by iamandrewluca** [comment]
@fregante are you ok with this? Should I add inversion of control to all functions?

**Comment by fregante** [comment]
The ideal way to support this would be to support plain strings as parameter, but that complicates things and forces each function to parse it every time. So your current solution of accepting a `URL` is interesting. 

I think it’s fair, however the default should be `url: URL | Location = location` since `location` is already `URL`-like and doesn’t need to be reparsed every time.

**Comment by iamandrewluca** [comment]
> however the default should be url: URL | Location = location since location is already URL-like and doesn’t need to be reparsed every time.

Ok. I agree with this. As a note, they differ a little bit

https://www.typescriptlang.org/play/index.html?ssl=1&ssc=1&pln=2&pc=61#code/MYewdgzgLgBAqgJwDYHUCWUAWICuUAyIwAhlGuAFwwCiAHsEjgCYCmAPANYsCeIAZvABK+ADQwuvAYRJlwAPgBQoSLGmlyYdFlxRESKnQbN2E-jDWywY0wLjC5QA

After first iteration, I thought maybe functions should receive a object, so more options can be passed. Exact types can be infered from used helpers parameters

e.g.

```ts
type Defaults = {
	url: Location | URL
	title: string
}

export const getDefaults = (): Defaults => ({
	url: location,
	title: document.title,
});

export const is404 = (o = getDefaults()): boolean => o.title === 'Page not found · GitHub';

type IsBlameOptions =
  & Parameters<typeof getRepoPath>
  & {} // ...
export const isBlame = (o: IsBlameOptions = getDefaults()): boolean => String(getRepoPath(o)).startsWith('blame/');
```

**Comment by fregante** [comment]
No, let’s leave DOM-based tests out of this. **In a separate PR**, maybe, those functions can accept a `document` object, which also has `document.location`

**Comment by iamandrewluca** [comment]
I'm thinking of cases where you need both url and title.

**Comment by iamandrewluca** [comment]
I updated to use `url: URL | Location = location` on `isRepoDiscussionList` and all nested calls

**Comment by fregante** [comment]
Looks good. Can you update all the URL-based functions? These are only 10 out of ~50

**Comment by iamandrewluca** [comment]
Ok.

**Comment by fregante** [comment]
I think this is also an opportunity to drop the `domTestsIdentifiers` regex in `test.ts` since DOM-based functions will now have a distinctive signature:

```diff
- 	if (validURLs === 'combinedTestOnly' || domTestsIdentifiers.test(detect.toString())) {
+ 	if (validURLs === 'combinedTestOnly' || detect.toString().startsWith('() =>') {
```

And in the future this would be changed to 


```diff
+ 	if (validURLs === 'combinedTestOnly' || detect.toString().startsWith('(document_ = document) =>') {
```

**Comment by iamandrewluca** [comment]
@fregante done. It seems they are more than 10 :slightly_smiling_face:

**Comment by fregante** [comment]
Don't worry about rebasing the PR, it will be squashed anyway

**Comment by iamandrewluca** [comment]
@fregante `domTestsIdentifiers` change, breakes tests.

**Comment by fregante** [comment]
It's failing because those 3 functions are not URL-based but DOM-based. 2 use `isUserProfile` and one uses `exists`. You need to revert the changes to those 3 functions.

File: source/index.ts [hunk 1]
```diff
@@ -293,108 +293,108 @@ collect.set('isRepoRoot', [
...
+export const isRepoTree = (url: URL | Location = location): boolean => isRepoRoot(url) || String(getRepoPath(url)).startsWith('tree/');
 collect.set('isRepoTree', [
 	...collect.get('isRepoRoot') as string[],
 	'https://github.com/sindresorhus/refined-github/tree/master/distribution',
 	'https://github.com/sindresorhus/refined-github/tree/0.13.0/distribution',
 	'https://github.com/sindresorhus/refined-github/tree/57bf435ee12d14b482df0bbd88013a2814c7512e/distribution',
 ]);
 
-export const isRepoWithAccess = (): boolean => isRepo() && exists('.reponav-item[href$="/settings"]');
+export const isRepoWithAccess = (url: URL | Location = location): boolean => isRepo(url) && exists('.reponav-item[href$="/settings"]');
```
Comment (fregante) [inline source/index.ts:317]:
```suggestion
export const isRepoWithAccess = (): boolean => isRepo() && exists('.reponav-item[href$="/settings"]');
```

File: source/index.ts [hunk 2]
```diff
@@ -91,69 +91,69 @@ collect.set('isGlobalDiscussionList', [
...
-export const isOrganizationDiscussion = (): boolean => /^orgs\/[^/]+\/teams\/[^/]+($|\/discussions)/.test(getCleanPathname());
+export const isOrganizationDiscussion = (url: URL | Location = location): boolean => /^orgs\/[^/]+\/teams\/[^/]+($|\/discussions)/.test(getCleanPathname(url));
 collect.set('isOrganizationDiscussion', [
 	'https://github.com/orgs/refined-github/teams/core-team/discussions?pinned=1',
 	'https://github.com/orgs/refined-github/teams/core-team/discussions/1',
 	'https://github.com/orgs/refined-github/teams/core-team',
 ]);
 
-export const isOwnUserProfile = (): boolean => getCleanPathname() === getUsername();
+export const isOwnUserProfile = (url: URL | Location = location): boolean => getCleanPathname(url) === getUsername();
```
Comment (fregante) [inline source/index.ts:146]:
```suggestion
export const isOwnUserProfile = (): boolean => getCleanPathname() === getUsername();
```

File: source/index.ts [hunk 3]
```diff
@@ -293,108 +293,108 @@ collect.set('isRepoRoot', [
...
-export const isBranches = (): boolean => String(getRepoPath()).startsWith('branches') ?? false;
+export const isBranches = (url: URL | Location = location): boolean => String(getRepoPath(url)).startsWith('branches') ?? false;
 collect.set('isBranches', [
 	'https://github.com/sindresorhus/refined-github/branches',
 ]);
 
 export const isUserProfile = (): boolean => exists('.user-profile-nav');
 
-export const isUserProfileRepoTab = (): boolean =>
+export const isUserProfileRepoTab = (url: URL | Location = location): boolean =>
```
Comment (fregante) [inline source/index.ts:358]:
```suggestion
export const isUserProfileRepoTab = (): boolean =>
```

File: source/index.ts [hunk 4]
```diff
@@ -293,108 +293,108 @@ collect.set('isRepoRoot', [
...
 	isUserProfile() &&
-	new URLSearchParams(location.search).get('tab') === 'repositories';
+	new URLSearchParams(url.search).get('tab') === 'repositories';
```
Comment (fregante) [inline source/index.ts:360]:
```suggestion
	new URLSearchParams(location.search).get('tab') === 'repositories';
```

**Comment by fregante** [comment]
Published as 1.1.0. Thank you!

If needed, feel free to add a `document_ = document` argument to DOM-based detections as well

**Pull request merged**